### PR TITLE
refactored take attempts to include process_response

### PR DESCRIPTION
**Pull request by mkatychev**
* request retry attempts now include a `process_response` comparison
* `ToTakeHiddenColouredJson` is now a generic trait
* `ToStringHidden` is now a generc trait
* moved styler out of take.rs and into lib.rs

### CODE REVIEW

File: src/take.rs [hunk 1]
```diff
@@ -21,60 +23,108 @@ use std::{
...
+    let payload_matches = match frame
+        .response
+        .match_payload_response(&frame.cut, &payload_response)
+    {
+        Err(e) => {
+            log_mismatch(&frame.response, &payload_response).context("fn log_mismatch failure")?;
+            return Err(Error::from(e));
+        }
+        Ok(r) => r,
+    };
```
Comment (mplanchard) [inline src/take.rs:53]: nit: this `match` could be replaced with `.map_err(|e| Error::from).or_else(|e| { log_mismatch(...); e })?`

Comment (mkatychev) [reply src/take.rs:53]: good idea, I need to get into the habit of using `map_err`

**Comment by mkatychev** [comment]
> Changes look good to me. Might be worth eventually considering making a struct to hold the parameters and making a trait to specify the `request:` function signature if you find yourself needing to use these in a generic context

I could always add a `request: fn(Params, Request) -> Result<Response, Error>` field to `Params` and populate it when I call `base_params.init()` but then I'd have to make `request` a method... since I can't pass it into its own property 🤔 
https://github.com/Bestowinc/darkroom/blob/4d603a415536527e4684316df9d0649f4e4a9be3/src/params.rs#L8-L15

**Comment by mplanchard** [comment]
> I could always add a request: fn(Params, Request) -> Result<Response, Error> field to Params and populate it when I call base_params.init() but then I'd have to make request a method... since I can't pass it into its own property 🤔

I'm not sure I have all the context needed to be useful here, and a refactor is almost certainly not worth it, but I sat noodling for a bit for fun (clearly I have nothing better to do 🙄), and this is the best I could come up with:

```rust
trait Protocol {
  fn request(prm: Params, req: Request) -> Result<Response, Error>;
}

struct Http;
impl Protocol for Http {
  fn request(prm: Params, req: Request) -> Result<Response, Error> {
    ...
  }
}

struct Grpc;
impl Protocol for Grpc {
  fn request(prm: Params, req: Request) -> Result<Response, Error> {
    ...
  }
}


struct Frame<T: Protocol> {
  protocol: T
}
impl<T: Protocol> Frame<T> {
  fn request(self, prm: Params) -> Result<Response, Error> {
    T::request(prm, self.get_request())
  }
}
```

The generic type would get associated w/the frame when it's constructed, and then `frame.request` would automatically call the right protocol's `request` method based on the generic type of the frame. Here's a [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5552b62ca8e23e69f36c7253b9460827) with a functional example.

if you're not actually referencing `protocol` anywhere outside of using to select a `request` method, you could wrap it in a `PhantomData<T>` marker so the compiler won't warn about it being unused.

**Comment by mplanchard** [comment]
P.S. I had way too much fun thinking about that. So glad you're working on Rust stuff!

**Pull request merged**